build(aio) + docs: miscellaneous updates to the doc-gen

petebacondarwin opened the pull request:
There are quite a lot of commits here but each is fairly self-contained and intuitive.

petebacondarwin committed:
build(aio): don't render `@Annotation` tags

Because we were "ignoring" these tags they were being
rendered as part of the previous tag.
What we really want to do is know about them, so that we
don't break the doc-gen but then ignore them when rendering.

petebacondarwin committed:
build(aio): do not try to auto-link to internal API items

This would cause dangling links since the target, being internal,
would not exist in the docs.

petebacondarwin committed:
build(aio): don't constrain `checkContentRules` to run before another processor

We don't really care when this processor runs as long as it happens
after the tags have been extracted.
By not constraining its `runBefore` property we can ensure that other
processors can be run before it.

petebacondarwin committed:
build(aio): add content rule to prevent usageNotes in non-export API docs

This commit also factors out `API_CONTAINED_DOC_TYPES` to be used by
both `filterContainedDocs` and `addAllowedPropertiesRules`.

petebacondarwin committed:
build(aio): update to latest dgeni-packages@0.26.2

This update allows us to autolink to methods and properties, which
means that we can change things like `{@link transition transition()}`
to just `transition()`.

petebacondarwin committed:
build(aio): do not allow `@usageNotes` on properties

petebacondarwin committed:
build(aio): refactor the decorator doc processing

petebacondarwin committed:
docs(aio): remove unused guide doc

This was erroneously committed into master, when it was really only
supposed to demo what the pages might look like.

petebacondarwin committed:
build(aio): improve automatic linking of code items

This commit adds new link disambiguators that mean that more
code links can be generated automatically in a sensible way.

The best example is the use of properties within class, interface and
enum documentation.

petebacondarwin committed:
docs: remove unnecessary `@linkDocs` tags

It is cleaner and simpler to use just a straightforward link.

petebacondarwin committed:
build(aio): remove unused `@linkDocs` alias for `@link` jsdoc tag

petebacondarwin committed:
build(aio): rearrange decorator API doc template

The overview of the decorator options is now a table.
The detailed description of each option is now a full section.

petebacondarwin committed:
build(aio): map H3 headings into H4 headings for certain templates

The sections such as methods and decorator options are already headed
by a H3 heading so we need to map the H3 headings in the API doc source
down to H4 headings.

This commit includes general heading mapping functionality accessible via
the `marked` Nunjucks filter.

petebacondarwin committed:
docs: fix invalid headings

petebacondarwin committed:
build(aio): rearrange processors to ensure we catch all content errors

mary-poppins commented:
You can preview af721ca at https://pr24000-af721ca.ngbuilds.io/.

gkalpak commented on aio/tools/transforms/templates/api/includes/decorator-overview.html:
```diff
@@ -1,10 +1,33 @@
+{% import "lib/descendants.html" as descendants -%}
+
+{% macro renderOptionsTable(doc) %}
+<table class="is-full-width list-table options-table">
```
1. Is the indentation here intentional?
2. Should `options-table` be `option-table`?

petebacondarwin replied:
1. No the indentation is totally screwy
2. yes

gkalpak commented on aio/tools/transforms/templates/api/includes/decorator-overview.html:
```diff
@@ -1,10 +1,33 @@
+          </a>
+        </td>
+        <td>{$ option.shortDescription | marked $}</td>
+    </tr>
```
Incorrect indentation.

gkalpak commented on aio/tools/transforms/templates/api/decorator.template.html:
```diff
@@ -5,5 +5,40 @@
+  <h2>Options</h2>
+  {% for option in doc.members %}
+  <a id="{$ option.anchor $}"></a>
+  <table class="is-full-width option-table">
```
Incorrect indentation 😱

gkalpak replied:
Indentation still incorrect :grin:

petebacondarwin replied:
I have unindented the `section` tag which should now solve this problem.

gkalpak commented on aio/tools/transforms/templates/api/decorator.template.html:
```diff
@@ -5,5 +5,40 @@
+  {% for option in doc.members %}
+  <a id="{$ option.anchor $}"></a>
+  <table class="is-full-width option-table">
+      <thead><tr><th><h3>{$ option.name $}</h3></th></tr></thead>
```
Is the `h3` needed here?

petebacondarwin replied:
The headings for the properties table also uses H3 here.
I guess we could avoid both and use appropriate changes to CSS.

petebacondarwin replied:
Actually thinking about this again, I think we should keep H3s here. Although wrapped in a table (for rendering) the semantic meaning is that the name of the option/property is the heading for this section, which describes this option/property.

gkalpak commented on aio/tools/transforms/templates/api/includes/decorator-overview.html:
```diff
@@ -1,10 +1,33 @@
 {% if doc.members.length %}
 <section class="decorator-overview">
-<code-example language="ts" hideCopy="true">
-@{$ doc.name $}{$ doc.typeParams | escape $}({ {$ memberHelper.renderMembers(doc) $}
```
Isn't this needed any more?

gkalpak commented on aio/tools/transforms/templates/api/lib/memberHelpers.html:
```diff
@@ -113,12 +113,24 @@ <h4 class="no-anchor">Throws</h4>
+  {% if method.usageNotes -%}
+    <tr>
+      <td class="usage-notes">
+        <h3>Usage Notes</h3>
```
Does this need the `.no-anchor` class (as in [decorator.template.html][1])?

[1]: https://github.com/angular/angular/pull/24000/files#diff-3ab70739108ff0d7be74c2f43afca985R35

petebacondarwin replied:
Actually I think that Usage Notes here should perhaps be a H4 since the main heading for this table is the H3 property name.

petebacondarwin replied:
Even further I decided that there should be no heading at all in this case!

petebacondarwin replied:
And the same for the decorator case.

gkalpak reviewed (changes requested):
Good stuff 💯
(Only some minor comments.)

petebacondarwin commented on aio/tools/transforms/angular-api-package/index.js:
```diff
@@ -177,3 +169,45 @@ module.exports = new Package('angular-api', [basePackage, typeScriptPackage])
+}
+
+function isMethod(doc) {
+  return doc.hasOwnProperty('parameters') && !doc.isGetAccessor && !doc.isSetAccessor;
```
I think so yes.

petebacondarwin commented on aio/tools/transforms/angular-api-package/index.js:
```diff
@@ -177,3 +169,45 @@ module.exports = new Package('angular-api', [basePackage, typeScriptPackage])
+    const ruleSet = checkContentRules.docTypeRules[docType] = checkContentRules.docTypeRules[docType] || {};
+
+    const rules = ruleSet['usageNotes'] = ruleSet['usageNotes'] || [];
+    rules.push((doc, prop, value) => value && !isMethod(doc) &&
```
Not quite. This rule only runs on docs that are of the types given in `API_CONTAINED_DOC_TYPES`. It just turns out that all methods and properties are of the same `docType` (i.e. `member`) so we have to have this additional check here.
So since the rule doesn't run on other docTypes (e.g. `class` or `function`) they are not constrained by it.

gkalpak replied:
So, this means that `usageNotes` are not allowed in members, except for methods (where they are allowed)?

petebacondarwin committed:
build(aio): remove invalid H3 usage notes heading

This heading is too high for the section because the
method name is a H3 but it cannot be a H4 because
usage notes may contain H4 headings.

petebacondarwin commented:
OK I fixed up commits based on all those lovely comments. Thanks @gkalpak 
I have now noticed two new issues:

* we are capturing the docs on the  decorator `const` declaration, which is not really desirable
* "Query" (e.g. `ViewChild`) decorator options are missing. I think this is because these are declared in a more complicated manner than most other decorators.

I suggest that we get this merged and fix those in subsequent PRs.

petebacondarwin commented on packages/core/src/di/forward_ref.ts:
```diff
@@ -44,11 +45,12 @@ export function forwardRef(forwardRefFn: ForwardRefFn): Type<any> {
  * {@example core/di/ts/forward_ref/forward_ref_spec.ts region='resolve_forward_ref'}
  *
- * See: {@link forwardRef}
+ * @see `forwardRef`
```
I have updated the others to use `@see` as well.

mary-poppins commented:
You can preview 704ab82 at https://pr24000-704ab82.ngbuilds.io/.

mary-poppins commented:
You can preview d6d73eb at https://pr24000-d6d73eb.ngbuilds.io/.

IgorMinar commented:
- [ ] core/Directives
- [ ] breadcrumbs missing
- [ ] edit & source code button for individual options (ideally this should also show up for class methods and anything similar to that, but it's ok if this is a separate PR)

petebacondarwin commented:
* [ ] fail doc-gen if invalid tag is found (requires change to dgeni-packages)

petebacondarwin committed:
build(aio): fix broken doc-gen unit test

petebacondarwin committed:
fix(aio): tidy up API doc styles

* Code anchors should inherit the font size from their container
* Table headings should align with content

petebacondarwin committed:
build(aio): fix decorator doc "inherited from" heading

We should not include the package path in the inherited
from heading for decorator API docs

petebacondarwin committed:
build(aio): compute breadcrums for all API doc types

petebacondarwin committed:
build(aio): add github links to API doc members

This change adds Github edit and view links to methods
and decorator options.

It is possible to add these to properties also but the
UI is rather tight as these are displayed in a table.

petebacondarwin commented:
I have cleaned up this PR and added commits for the items that Igor requested above.

mary-poppins commented:
You can preview 9842371 at https://pr24000-9842371.ngbuilds.io/.

gkalpak commented on aio/src/styles/1-layouts/_api-page.scss:
```diff
@@ -14,7 +14,7 @@
             border-radius: 4px 4px 0 0;
             text-transform: none;
-            padding: 8px 24px;
+            padding: 8px 16px;
```
Are `.api-section` and `.api-table` used anywhere? I couldn't find any occurrences :confused:

petebacondarwin replied:
weird me neither...

petebacondarwin replied:
I'm going to delete this and consolidate the other styles into the `styles/2-modules/_api-pages.scss` file

gkalpak commented on aio/tools/transforms/angular-api-package/processors/computeApiBreadCrumbs.js:
```diff
@@ -1,16 +1,15 @@
+        if (API_DOC_TYPES_TO_RENDER.indexOf(doc.docType) !== -1) {
+          doc.breadCrumbs = [];
+          doc.breadCrumbs.push({ text: 'API', path: '/api' });
+          if (doc.moduleDoc) doc.breadCrumbs.push({ text: '@angular/' + doc.moduleDoc.id, path: doc.moduleDoc.path });
```
1. There is no test for this conditional 😱
2. Under what circumstances can a `doc` have no corresponding `moduleDoc`? Do you have an example?

petebacondarwin replied:
when it is a module? I'll add a test

gkalpak commented on aio/tools/transforms/templates/api/lib/githubLinks.html:
```diff
@@ -9,4 +9,11 @@
+{%- endmacro -%}
+
+{% macro githubLinks(doc, versionInfo) -%}
+<div class="page-actions">
```
`page-actions` might not be an descriptive name any more :grin:

petebacondarwin replied:
renamed to `github-links`

gkalpak reviewed (approved):
Excited to have the `view`/`edit` buttons on members \o/
Not thrilled about how they look (and adding red arrows didn't actually help) :grin: :

![view-edit-members](https://user-images.githubusercontent.com/8604205/40473474-f7c6587e-5f44-11e8-9ba9-d08e42030854.png)

mary-poppins commented:
You can preview c065871 at https://pr24000-c065871.ngbuilds.io/.

petebacondarwin commented:
@gkalpak do you have any suggestions for better styling of the GH links on methods and options?

gkalpak commented:
I'm no designer 😛

But I would definitely start at centering them vertically:
![member-gh-links-1](https://user-images.githubusercontent.com/8604205/40539310-04265a50-601d-11e8-94fa-efdb0b5f9e54.png)

And I would try out different colors (a dark-ish gray maybe, to make them stand out less):
![member-gh-links-2](https://user-images.githubusercontent.com/8604205/40539308-03f621c8-601d-11e8-8d18-5bc2ce140e5a.png)

@sjtrimble will definitely have more useful insights than me :grin:

sjtrimble commented:
Coming late to the party - is the thought that the main page source and edit would still be there in addition to the 'individual option' ones?

I like @gkalpak idea of having the buttons be more subtle. Would be nice if they turned blue on hover (but without the rounded box border of the current page icons).

Aligning the card header content vertically is as intended in the original design - so that's a go from my perspective :)

petebacondarwin commented:
@sjtrimble:

> is the thought that the main page source and edit would still be there in addition to the 'individual option' ones?

Yes that is correct. The "individual option" ones link to the source for those specific items (e.g. the method), whereas the top level on is for the export as a whole (e.g. the class).

I have centred and coloured the links as suggested.

PTAL

mary-poppins commented:
You can preview 1b5959c at https://pr24000-1b5959c.ngbuilds.io/.

petebacondarwin commented:
<img width="1051" alt="screen shot 2018-06-04 at 15 40 53" src="https://user-images.githubusercontent.com/15655/40923825-b585566e-680d-11e8-83f0-c49c42330bd3.png">

This shows one link being hovered (in blue) while the other is normal (grey).

jbogarthyde commented:
I can't see where one overload ends and another one starts -- I would like to see at least an <hr> between them - preferably a descriptive header. How do you navigate to the one you want among 17 almost identical versions?
I have some test doc for the HttpClient file in another PR -- https://github.com/angular/angular/pull/23839

petebacondarwin commented:
@jbogarthyde there actually are `<hr>`s between the overloads but they are quite faint! I think the main problem is that the "heading" of each overload is being rendered in a non-bold font and also has line numbers, which makes them look like an example rather than a heading.

Currently we don't have an anchor on each of the overloads so you cannot navigate to them directly. Moreover since these overloads are hidden inside a closed `<details>` element then I am not sure if they would automatically open and scroll to this position even if you could link to them. Is that what you mean by navigating to them?

mary-poppins commented:
You can preview 7c06727 at https://pr24000-7c06727.ngbuilds.io/.

mary-poppins commented:
You can preview 9697d53 at https://pr24000-9697d53.ngbuilds.io/.

jbogarthyde commented:
It would be nice to have a descriptive header that made it into the right-hand TOC, but failing that, you should be able to scroll through opened detail pane and pick out the one you want. 
I guess what I would like to see (in addition to removing the line numbers) is a header between each @returns line and the next signature, that describes the difference -- something like "Get a blog", "Get an observable", "Get observables"...  The header should stand out a bit from the param/return headers, maybe slightly bigger, different color?

petebacondarwin commented:
@jbogarthyde can you create a new issue to track this?

IgorMinar commented:
@petebacondarwin it looks like this PR breaks CI. can you please check? thanks

petebacondarwin commented:
The breakage was not due to a direct change. So I am rebasing again to see if that makes things better.

mary-poppins commented:
You can preview 48cfabe at https://pr24000-48cfabe.ngbuilds.io/.

petebacondarwin commented:
All green!

mhevery commented:
This PR can't be merged because the `fixup!` do not apply cleanly. Could you squash the `fixups` into the corresponding SHAs and than it can be merged.

mary-poppins commented:
You can preview 14acfc2 at https://pr24000-14acfc2.ngbuilds.io/.
